src/network/serverpackethandler.cpp

```diff
@@ -1339,10 +1339,37 @@ void Server::handleCommand_RemovedSounds(NetworkPacket* pkt)
 		if (i == m_playing_sounds.end())
 			continue;
 
+		session_t peer_id = pkt->getPeerId();
+
 		ServerPlayingSound &psound = i->second;
-		psound.clients.erase(pkt->getPeerId());
-		if (psound.clients.empty())
+		psound.clients.erase(peer_id);
+			verbosestream << "Server: Sound " << id << " erase peer "
+					<< peer_id << "." << std::endl;
+		if (!psound.can_be_send_later && psound.clients.empty()) {
 			m_playing_sounds.erase(i);
+			verbosestream << "Server: Erase sound " << id << "." << std::endl;
+		}
+		else if (psound.can_be_send_later && !psound.spec.loop) {
+			if (psound.keep_time < m_playing_sounds_time) {
+				/* Sound keep time reached, done_clients do not need to be updated. */
+				verbosestream << "Server: Done sound " << id << " for peer "
+						<< peer_id << "." << std::endl;
+				continue;
+			}
+
+			PlayerSAO *plrsao = getPlayerSAO(peer_id);
+			if (!plrsao)
+				continue;
+			v3f pos = psound.getPos(m_env, nullptr);
+
+			if (pos.getDistanceFrom(plrsao->getBasePosition()) <= psound.max_hear_distance) {
+				/* If client remove sound in hear distance,
+				   we are sure, that sound has been fully played. */
+				psound.done_clients.insert(peer_id);
```

Comment:
This logic seems brittle to me: The positions on server and client may not be perfectly in sync, for example due to attached sounds, meaning that this check can't accurately match what the client is doing.

The straightforward fix might be to have the client additionally tell us why a sound was removed (whether it was removed for going out of range or for reaching the end of playback) by extending the packet. For older clients, we could still keep this hack as a decent approximation.

Reply:
There is no packet sent when the client comes out of the sound range. The Sound continues playing.
Clients generate stop sound packets only if the playing sound is finished on the client side. When sound is removed due to server request, no stop sound packet is generated.